---
title: "Symbiodinium analysis of corals from Hurricane Hole"
subtitle: ""
author: "Ross Cunning"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath(".."))
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
options(scipen = 3, digits = 9)
```

```{r setup_2, include=FALSE}
# Load custom R functions
source("R/functions.R", .GlobalEnv)
# Load package libraries
library(phyloseq); library(vegan); library(multcompView); library(reshape2); library(igraph); library(stringr); library(tidyverse); library(cowplot)
library(ForceAtlas2) # devtools::install_github("analyxcompany/ForceAtlas2")
# Set colors for plotting clades
taxcolors <- matrix(c("#8dd3c7", "#ffffb3", "#bebada", "#fb8072", "#80b1d3"), 
                    dimnames=list(c("CladeA", "CladeB", "CladeC", "CladeD", "CladeG")))
```

# Project
  This project investigates *Symbiodinium* communities in three species of corals: *Colphophyllia natans*, *Orbicella annularis*, and *Diploria labyrinthiformis*, at sites inside Hurricane Hole (Otter Creek and Water Creek) and outside (Lime Tree, Coral Bay, Haulover), in St. John, USVI. Bacterial communities from the same samples were analyzed by Erinn Muller. 
  The major aims of this work are to:
  
1. Characterize *Symbiodinium* communities in these corals and test for differences inside vs. outside Hurricane Hole.
2. Pair the *Symbiodinium* data with bacterial community data to investigate links between these compartments of the microbiome.

# Data

Read in phyloseq object generated by SymITS2 scripts. This pipeline uses 97% clustering within samples to generate OTUs, and global alignment to a custom reference database to assign taxonomy. Sequences with poor matches to the *Symbiodinium* database are compared to the NCBI nt database and removed if they match well to a non-*Symbiodinium* sequence.

```{r load_data}
# Read in phyloseq object
load("data/phy.f.RData")

# Filter OTUs by minimum count
# Set threshold count
n <- 10
# Identify OTUs below threshold count
lowcount_taxa <- taxa_sums(phy.f)[which(taxa_sums(phy.f) >= n)]
# Remove taxa below threshold count
phy.f <- prune_taxa(names(lowcount_taxa), phy.f)
phy.f <- filter_taxa(phy.f, function(x) sum(x >= 3) >= 3, TRUE)

# Label clades and subtypes for filtered phyloseq object tax_tables
get.st <- function(df) {
  within(df, {
    Clade <- substr(hit, 1, 1)
    Subtype <- gsub(hit, pattern="_[A-Z]{2}[0-9]{6}", replacement="")
    Subtype <- gsub(Subtype, pattern="_multiple", replacement="")
    Subtype2 <- ifelse(as.numeric(sim)==100, paste0("'", Subtype, "'"),
                       paste0("'", rep(rle(sort(Subtype))$values, times=rle(sort(Subtype))$lengths), "'^", 
                              unlist(lapply(rle(sort(Subtype))$lengths, seq_len)))[order(order(Subtype))])
    #Subtype <- ifelse(as.numeric(sim)==100, Subtype, paste("*", Subtype, sep=""))
  })
}

tax_table(phy.f) <- as.matrix(get.st(data.frame(tax_table(phy.f), stringsAsFactors=FALSE)))

```

### Descriptive statistics

```{r phy.f_histograms, cache=T, fig.height=4, fig.width=4}
# Compute summary statistics
stats <- data.frame(value=t(phystats(phy.f)), check.names=F)

# Create and plot histograms
taxhist <- hist(log10(taxa_sums(phy.f)), plot=F)
samhist <- hist(log10(sample_sums(phy.f)), plot=F)

par(mfrow=c(2, 1), mar=c(3,3,1,1))
plot(taxhist, col="black", main="97% OTU counts", xlim=c(0, 6), las=1, mgp=c(2,0.5,0),
     xlab="No. sequences (log10)", ylab="No. OTUs", cex.lab=0.75, cex.axis=0.75)
plot(samhist, col="black", main="97% OTU reads per sample", xlim=c(0, 6), las=1, mgp=c(2,0.5,0),
     xlab="No. sequences (log10)", ylab="No. samples", cex.lab=0.75, cex.axis=0.75)

# Create stats table
knitr::kable(stats)
```

### Transform count data
Count data are transformed to both relative abundance (proportions) and square-root proportions for downstream statistical analyses.
```{r transform}
# Convert to proportion (relative abundance)
phy.f.p <- transform_sample_counts(phy.f, function(x) x/sum(x))
# Apply transformation function
transform <- function(x) sqrt(x/sum(x))  # Set transformation function
phy.f.t <- transform_sample_counts(phy.f, transform)  # Transform data
save(phy.f.p, file = "data/phy.f.p.RData")
```

# Results

## Overview

Composition of *Symbiodinium* clades in each sample of each species at each site.

```{r, fig.height=10, fig.width=10}
# Reduce taxonomy table to Clade > Subtype > Subtype2
tax2 <- select(data.frame(tax_table(phy.f.p)), Clade, Subtype, Subtype2)
tax_table(phy.f.p) <- tax_table(as.matrix(tax2))

clades <- phy.f.p %>%
  tax_glom(taxrank = "Clade") %>%                # agglomerate at clade level
  psmelt() %>%                                         # Melt to long format
  arrange(Clade)                                      # Sort data frame alphabetically by phylum

plot_clades <- function(data, title) {
  ggplot(data, aes(x = Sample, y = Abundance, fill = Clade)) +
  facet_grid(~Site, scales = "free", drop = FALSE) +
  geom_bar(stat = "identity") +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 1)) +
  scale_fill_manual(values = taxcolors) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        plot.title = element_text(hjust = 0, size=16)) +
  labs(title=title, x = NULL, y = "Relative abundance")
}

cnat <- plot_clades(data = filter(clades, Species=="Cnat"), title = "C. natans")
dlab <- plot_clades(data = filter(clades, Species=="Dlab"), title = "D. labyrinthiformis")
oann <- plot_clades(data = filter(clades, Species=="Oann"), title = "O. annularis")

plot_grid(oann, dlab, cnat, nrow=3, ncol=1)
```

----------

## *Diploria labyrinthiformis*

#### Barplot
```{r dlab_bars, eval=T, cache=F, fig.height=3, fig.width=8}
dlab.phy <- phy.f.p %>%
  subset_samples(., Species=="Dlab") %>%                  # subset D. lab samples
  filter_taxa(., function(x) sum(x >= 0.01) >= 1, TRUE)   # subset taxa >=1% in at least 1 sample

plot_bar(physeq=dlab.phy, fill = "Subtype2", title = "D. labyrinthiformis") +
  facet_grid(~Site, scales = "free_x")
```

#### NMDS
```{r dlab_nmds}
# D. labyrinthiformis NMDS
dlab.phy <- subset_samples(phy.f.t, Species=="Dlab")      # use sqrt-transformed data
set.seed(67827)
dlab.nmds <- phyloseq::ordinate(dlab.phy, method="NMDS", distance="bray", trace=0)
plot_ordination(physeq = dlab.phy, ordination = dlab.nmds, type = "samples", 
                color = "Site", shape = "Site")
```

#### Network
```{r dlab_net}
# D. labyrinthiformis network analysis
dlab.phy <- subset_samples(phy.f.p, Species=="Dlab")
dlab.net <- makenet(dlab.phy, thresh = 0.01, type = "igraph")

set.seed(2578)
fa2 <- layout.forceatlas2(dlab.net, iterations = 1000, gravity = 40, tolerance = 10,
                          plotstep = FALSE)
par("mar"=c(0,0,0,0))
plot(dlab.net, layout=fa2,
     edge.curved=0.1, 
     vertex.label.cex=1,
     edge.width=3*(E(dlab.net)$weight)^0.35,
     vertex.label=parse(text=ifelse(is.na(V(dlab.net)$Clade),
                                    substr(names(V(dlab.net)), 1, 1), 
                                    V(dlab.net)$Subtype2)),
     vertex.color=ifelse(is.na(V(dlab.net)$Clade), "white", 
                         taxcolors[factor(V(dlab.net)$Clade, levels=c("A","B","C","D","G"))]),
     vertex.size=ifelse(is.na(V(dlab.net)$Clade), 5, 3*igraph::degree(dlab.net)^0.5),
     vertex.shape=c("none", "circle")[factor(is.na(V(dlab.net)$Site))])
```

## *Orbicella annularis*

#### Barplot
```{r oann, eval=T, cache=F, fig.height=3, fig.width=8}
oann.phy <- phy.f.p %>%
  subset_samples(., Species=="Oann") %>%                  # subset O. ann samples
  filter_taxa(., function(x) sum(x >= 0.01) >=1, TRUE)    # subset taxa >=1% in at least 1 sample

plot_bar(physeq=oann.phy, fill = "Subtype2", title = "O. annularis") +
  facet_grid(~Site, scales = "free_x")
```

#### NMDS
```{r oann_nmds}
# O. annularis NMDS
oann.phy <- subset_samples(phy.f.t, Species=="Oann")
set.seed(67827)
oann.nmds <- phyloseq::ordinate(oann.phy, method="NMDS", distance="bray", trace=0)
plot_ordination(physeq = oann.phy, ordination = oann.nmds, type = "samples", 
                color = "Site", shape = "Site")
```

#### Network
```{r oann_net}
# O. annularis network analysis
oann.phy <- subset_samples(phy.f.p, Species=="Oann")
oann.net <- makenet(oann.phy, thresh = 0.01, type = "igraph")

set.seed(2573)
fa2 <- layout.forceatlas2(oann.net, iterations = 6000, gravity = 2500, k=1e6, tolerance = 100,
                          plotstep = FALSE)
par("mar"=c(0,0,0,0))
plot(oann.net, layout=fa2,
     edge.curved=0.1, 
     vertex.label.cex=1,
     edge.width=3*(E(oann.net)$weight)^0.35,
     vertex.label=parse(text=ifelse(is.na(V(oann.net)$Clade),
                                    substr(names(V(oann.net)), 1, 1), 
                                    V(oann.net)$Subtype2)),
     vertex.color=ifelse(is.na(V(oann.net)$Clade), "white", 
                         taxcolors[factor(V(oann.net)$Clade, levels=c("A","B","C","D","G"))]),
     vertex.size=ifelse(is.na(V(oann.net)$Clade), 5, 3*igraph::degree(oann.net)^0.5),
     vertex.shape=c("none", "circle")[factor(is.na(V(oann.net)$Site))])
```

## *Colpophyllia natans*

#### Barplot
```{r cnat, eval=T, cache=F, fig.height=3, fig.width=8}
cnat.phy <- phy.f.p %>%
  subset_samples(., Species=="Cnat") %>%                  # subset C. nat samples
  filter_taxa(., function(x) sum(x >= 0.01) >=1, TRUE)    # subset taxa >=1% in at least 1 sample

plot_bar(physeq=cnat.phy, fill = "Subtype2", title = "C. natans") +
  facet_grid(~Site, scales = "free_x")
```

#### NMDS
```{r cnat_nmds}
# C. natans NMDS
cnat.phy <- subset_samples(phy.f.t, Species=="Cnat")
set.seed(67827)
cnat.nmds <- phyloseq::ordinate(cnat.phy, method="NMDS", distance="bray", trace=0)
plot_ordination(physeq = cnat.phy, ordination = cnat.nmds, type = "samples", 
                color = "Site", shape = "Site")
```

#### Network
```{r cnat_net}
# Network plot may be much better
cnat.phy <- subset_samples(phy.f.p, Species=="Cnat")
cnat.net <- makenet(cnat.phy, thresh = 0.01, type = "igraph")
set.seed(2558)
fa2 <- layout.forceatlas2(cnat.net, iterations = 1000, gravity = 100, tolerance = 10.0,
                          plotstep = FALSE)
par("mar"=c(0,0,0,0))
plot(cnat.net, layout=fa2,
     edge.curved=0.1, 
     vertex.label.cex=1,
     edge.width=3*(E(cnat.net)$weight)^0.35,
     vertex.label=parse(text=ifelse(is.na(V(cnat.net)$Clade),
                                    substr(names(V(cnat.net)), 1, 1), 
                                    V(cnat.net)$Subtype2)),
     vertex.color=ifelse(is.na(V(cnat.net)$Clade), "white", 
                         taxcolors[factor(V(cnat.net)$Clade, levels=c("A","B","C","D","G"))]),
     vertex.size=ifelse(is.na(V(cnat.net)$Clade), 5, 5*igraph::degree(cnat.net)^0.5),
     vertex.shape=c("none", "circle")[factor(is.na(V(cnat.net)$Site))])
```

